source/common/router/config_impl.h

```diff
@@ -344,6 +345,83 @@ class RouteTracingImpl : public RouteTracing {
   envoy::type::FractionalPercent overall_sampling_;
 };
 
+/**
+ * Base implementation for path rewriters
+ */
+class PathRewriter {
+public:
+  virtual bool apply() const = 0;
+  virtual std::string rewrite(std::string path, const std::string& matched_path,
+                              bool case_sensitive) const = 0;
+  virtual ~PathRewriter(){};
+};
+
+/**
+ * Path rewrite through prefix replacement
+ */
+class PrefixPathRewriter : public PathRewriter {
+public:
+  PrefixPathRewriter(std::string prefix) : PathRewriter(), prefix_(prefix) {}
+
+  bool apply() const { return !prefix_.empty(); }
+
+  std::string rewrite(std::string path, const std::string& matched_path,
+                      bool case_sensitive) const {
+    ASSERT(case_sensitive ? absl::StartsWith(path, matched_path)
+                          : absl::StartsWithIgnoreCase(path, matched_path));
+    return path.replace(0, matched_path.size(), prefix_);
+  }
+
+  ~PrefixPathRewriter(){};
+
+private:
+  std::string prefix_;
+};
+
+/**
+ * Path rewrite via regex
+ */
+class RegexPathRewriter : public PathRewriter {
+public:
+  RegexPathRewriter(std::string pattern, std::string substitution)
+      : PathRewriter(), pattern_(pattern), substitution_(substitution) {}
+
+  bool apply() const { return !substitution_.empty(); }
+
+  std::string rewrite(std::string path, const std::string& matched_path,
+                      bool case_sensitive) const {
+    ASSERT(case_sensitive ? absl::StartsWith(path, matched_path)
+                          : absl::StartsWithIgnoreCase(path, matched_path));
+    return std::regex_replace(matched_path, pattern_, substitution_);
+  }
+
+  ~RegexPathRewriter(){};
+
+private:
+  std::regex pattern_;
+  std::string substitution_;
+};
+
+/**
+ * Construct a path rewriter
+ */
+class RewriteBuilder {
+public:
+  static PathRewriter& build(const envoy::api::v2::route::Route& route) {
+    if (!route.redirect().prefix_rewrite().empty()) {
+      static PrefixPathRewriter p(route.redirect().prefix_rewrite());
+      return p;
+    }
+    if (!route.route().prefix_rewrite().empty()) {
+      static PrefixPathRewriter p(route.route().prefix_rewrite());
```

Comment:
The prefix is correctly returned from here according to the debug lines I added locally

```
[2019-10-02 16:40:30.377][14][debug][misc] [bazel-out/k8-fastbuild/bin/source/common/router/_virtual_includes/config_lib/common/router/config_impl.h:445] the RewriteBuilder prefix is: /api/new_endpoint
[2019-10-02 16:40:30.378][14][debug][misc] [bazel-out/k8-fastbuild/bin/source/common/router/_virtual_includes/config_lib/common/router/config_impl.h:445] the RewriteBuilder prefix is: /rewrote
[2019-10-02 16:40:30.379][14][debug][misc] [bazel-out/k8-fastbuild/bin/source/common/router/_virtual_includes/config_lib/common/router/config_impl.h:445] the RewriteBuilder prefix is: /rewrote
[2019-10-02 16:40:30.380][14][debug][misc] [bazel-out/k8-fastbuild/bin/source/common/router/_virtual_includes/config_lib/common/router/config_impl.h:445] the RewriteBuilder prefix is: /bar
[2019-10-02 16:40:30.380][14][debug][misc] [bazel-out/k8-fastbuild/bin/source/common/router/_virtual_includes/config_lib/common/router/config_impl.h:445] the RewriteBuilder prefix is: /car
[2019-10-02 16:40:30.380][14][debug][misc] [bazel-out/k8-fastbuild/bin/source/common/router/_virtual_includes/config_lib/common/router/config_impl.h:445] the RewriteBuilder prefix is: /cAndy
[2019-10-02 16:40:30.380][14][debug][misc] [bazel-out/k8-fastbuild/bin/source/common/router/_virtual_includes/config_lib/common/router/config_impl.h:445] the RewriteBuilder prefix is: /oranGES
```